src/gui/DatabaseTabWidget.cpp

```diff
@@ -363,6 +363,8 @@ bool DatabaseTabWidget::saveDatabase(Database* db)
             emit messageDismissTab();
             return true;
         } else {
+            dbStruct.modified = true;
```

Comment:
@droidmonkey the only problem I have with that is that if the database file was not modified but the user tried to save manually, and the save fails, the db will be marked as dirty even though no changes have been made. You can try for example by removing write permission to the db file after opening it in keepassxc. Of course you could argue that it's not gonna happen often.

What about marking the database dirty before calling autosave in `DatabaseTabWidget::modified()`:
```
diff --git a/src/gui/DatabaseTabWidget.cpp b/src/gui/DatabaseTabWidget.cpp
--- a/src/gui/DatabaseTabWidget.cpp
+++ b/src/gui/DatabaseTabWidget.cpp
@@ -780,16 +780,17 @@ void DatabaseTabWidget::modified()
     Database* db = static_cast<Database*>(sender());
     DatabaseManagerStruct& dbStruct = m_dbList[db];
 
-    if (config()->get("AutoSaveAfterEveryChange").toBool() && dbStruct.saveToFilename) {
-        saveDatabase(db);
-        return;
-    }
-
     if (!dbStruct.modified) {
         dbStruct.modified = true;
         dbStruct.dbWidget->databaseModified();
         updateTabName(db);
     }
+
+    if (config()->get("AutoSaveAfterEveryChange").toBool() && dbStruct.saveToFilename) {
+        saveDatabase(db);
+    }
+
 }
```

Reply:
The whole saving structure needs to be redone. I suggest moving this improvement edge case to then.

Reply:
@droidmonkey as you wish, but I thought the above diff was solving both problems.

Reply:
I will need to think about that a little bit. The case I am trying to solve with my fix is the user requested a save and it failed to save so I am marking the database as dirty because I do not know if the state can be trusted. This incidentally fixes the issue when an auto save fails and never marks the database dirty (ie. the database != save file).

Reply:
Is this ready to be merged?

Reply:
Yes. I will be refactoring the code above with my database pointer fixes and will address it there for 2.3